src/mainwindow.cpp

```diff
@@ -398,3 +488,80 @@ void MainWindow::aboutHotspot()
     dialog.adjustSize();
     dialog.exec();
 }
+
+void MainWindow::navigateToCode(const QUrl &url, int lineNumber, int columnNumber)
+{
+    QSettings settings(QStringLiteral("KDAB"), QStringLiteral("Hotspot"));
+    settings.beginGroup(QStringLiteral("CodeNavigation"));
+    const auto ideIdx = settings.value(QStringLiteral("IDE"), -1).toInt();
+
+    QString command;
+#if !defined(Q_OS_WIN) && !defined(Q_OS_OSX) // Remove this #if branch when adding real data to ideSettings for Windows/OSX.
+    if (ideIdx >= 0 && ideIdx < ideSettingsSize) {
+        command += QString::fromUtf8(ideSettings[ideIdx].app);
+        command += QString::fromUtf8(" ");
+        command += QString::fromUtf8(ideSettings[ideIdx].args);
+    } else
+#endif
+    if (ideIdx == -1) {
+        command = settings.value(QStringLiteral("CustomCommand")).toString();
+    } else {
+        QDesktopServices::openUrl(QUrl(url));
+    }
+
+    const QString filePath = url.isLocalFile() ? url.toLocalFile() : url.toString();
+    command.replace(QStringLiteral("%f"), filePath);
+    command.replace(QStringLiteral("%l"), QString::number(std::max(1, lineNumber)));
+    command.replace(QStringLiteral("%c"), QString::number(std::max(1, columnNumber)));
+
+    if (!command.isEmpty()) {
+        qDebug() << "Detaching: " << command;
+        QProcess::startDetached(command);
+    }
+}
+
+void MainWindow::setCodeNavigationIDE(QAction *action)
+{
+    QSettings settings(QStringLiteral("KDAB"), QStringLiteral("Hotspot"));
+    settings.beginGroup(QStringLiteral("CodeNavigation"));
+
+    if (action->data() == -1) {
+        const auto customCmd = QInputDialog::getText(
+            this, tr("Custom Code Navigation"),
+            tr("Specify command to use for code navigation, '%f' will be replaced by the file name, '%l' by the line number and '%c' by the column number."),
+            QLineEdit::Normal, settings.value(QStringLiteral("CustomCommand")).toString()
+            );
+        if (!customCmd.isEmpty()) {
+            settings.setValue(QStringLiteral("CustomCommand"), customCmd);
+            settings.setValue(QStringLiteral("IDE"), -1);
+        }
+        return;
+    }
+
+    const auto defaultIde = action->data().toInt();
+    settings.setValue(QStringLiteral("IDE"), defaultIde);
+}
+
+void MainWindow::onSourceMapContextMenu(const QPoint &point)
+{
+    const auto index = ui->sourceMapView->indexAt(point);
+    if (!index.isValid()) {
+        return;
+    }
+    const auto sourceMap = index.data(SourceMapModel::LocationRole).value<QString>();
+    auto seperator = sourceMap.lastIndexOf(QLatin1Char(':'));
+    if (seperator <= 0) {
```

Comment:
I wonder if we could ever get a situation where we know a filename, but no line number? in such a case, opening the file would still be useful. but probably it's OK to keep this as-is and if we ever see such a case we can improve its handling later on